Output to HDF5

benchislett opened the pull request:
# The Goal

Have file output be done in a way that is both compact, and easy to use.
Storing data as strings is suboptimal, and writing out many files per timestep is very inefficient.

# The Proposal

By using HDF5, we can efficiently store and even compress the data on disk.
Further, there are bindings for I/O in many popular languages, so ease of access will not be greatly reduced.

# The Method

At compile time, pull down hdf5 and install it automatically within the repo.
Then, link to the binaries with CMake.
Finally, rewrite the FileIO namespace to write all data into a single output file, which is closed when the program terminates.

# The Implementation

## CMake and Refactor

@mlxd has refactored the file structure and CMake build process in the branch `cmake_build`. These changes were merged into this branch very early on, as they were essential to the hdf5 build process.

## HDF5 Installation

The build script is a simple bash script stored in the `bin` folder that does the following:
- Exit if HDF5 is already installed here
- `wget` the HDF5 archive
- Extract the archive into the `bin` folder
- Configure and install HDF5 with the C++ bindings

The directory structure is directly linked in `CMakeLists.txt` and HDF5 is added with `find_package`

## FileIO Restructuring

### Output

Instead of writing out data with generically typed `writeOut{Int,Bool,Double,...}` functions, we create a generic helper function to write a dataset to the output file and wrap it in functions for each element we want to write.
i.e. `write1d` writes a generic dataset from contiguous memory, `writeNd` reshapes a vector of pointer arrays into a contiguous block and defers to `write1d`, and `writeOutWfc` and the like all wrap `writeNd`.

The structure of the output file is similar to that of a directory system, with Groups analogous to folders, and DataSets to files.

There are also attributes stored at root and in the `ENERGY` group, representing the `par` object and the energy at a given timestep respectively.

### Input

In order to continue simulations from an output file, we load the hfd5 data instead of creating it.
Groups specifically are loaded from the file initially, along with the `par` attributes, and the most recent `wfc`.
Later on, the `A` series (`Ax`, `Ay`, `Az`) are loaded when needed.

When writing to a DataSet, we now need a way to know if we should open from the file (when loading) or create a new one.
This is determined by checking the existence of the DataSet in the file, and conditionally loading either of them.
For simplicity, we also store an `unordered_map` of the DataSets so that we don't have to open/close them each time.

### Unchanged functionality

Some vortex manipulation still writes data directly to a file, untouched by this feature.
This is because of the dragons which guard the code, and the legacy functionality.
Only the edges are currently written to the hdf5 output file, though this may change in future releases.

## Additional Changes

In order for the loading to work correctly, some changes were made to the `Grid` class. Namely, we now store a default mapping for the default values of some parameters, which are checked iff the value does not exist in the central map.
This way, we can implement a `store_only` method which stores a value iff it has not previously been set.
With this, we can load the parameters from file after parsing without fear of overwriting explicitly set values.

 Also, `gstate` has been changed from an integer to a boolean, and the resulting fallout has been fixed so that the new representation (true=imaginary, false=real) is used.

A separate iteration counter for both evolution types has been added, which tracks the most recent iteration number for each evolution type.
This is used primarily for continuation of previous simulations via the `-r` flag, but is also a handy marker to know where we are in the simulation, and could be used in the future for monitoring execution and performance.
Because of this, and the new way of continuing simulations, `step_offset` has been removed entirely.

benchislett committed:
Add automatic compilation script

benchislett committed:
Update hdf5 build script to write to stdout

benchislett committed:
Merge branch 'development' into feature/output-to-hdf5

benchislett committed:
Use configure and make for install_hdf5 script

benchislett committed:
update makefile with hdf5 build script

mlxd committed:
Created libgpue_io.a structure

mlxd committed:
Moving vortex files into libs for lattice and vortices separately

mlxd committed:
Subdir including for diff sub-projects

mlxd committed:
Restructured for CMake build.

mlxd committed:
Renaming many files and locations

benchislett committed:
initialize hdf5 in fileio

benchislett committed:
update cmakelists with find hdf5 cmake build

benchislett committed:
fix some bugs with cmake_build

mlxd committed:
Merge pull request #40 from GPUE-group/cmake_changes

fix some bugs with cmake_build

benchislett committed:
Merge branch 'cmake_build' into feature/output-to-hdf5

benchislett committed:
Working CMake build for manual hdf5 build

benchislett committed:
Remove cuda_runtime

The <cuda_runtime.h> header was included, which corresponds to the C++ API described here:
https://www.cs.cmu.edu/afs/cs/academic/class/15668-s11/www/cuda-doc/html/group__CUDART__HIGHLEVEL.html#_details

From what I grepped, none of these functions are used, so I removed the inclusion altogether

benchislett committed:
Remove linking to simulator from tests,

This linking should already happen at a higher level

benchislett committed:
Link to object libraries with $<TARGET_OBJECTS>

benchislett committed:
Check if 3.12 policy exists before disabling

benchislett committed:
Add include directory to CMakeLists

benchislett committed:
Remove spacing causing bug in CMakeLists.txt

benchislett committed:
Rearrange elements in CMakeLists to compile correctly

benchislett committed:
Merge branch 'feature/output-to-hdf5' of https://github.com/GPUE-group/GPUE into feature/output-to-hdf5

benchislett committed:
Add initializer and destructor for hdf5 IO

benchislett committed:
Add hdf5 writeOut methods for Wfc, V, and K

benchislett committed:
Add edges to hdf5 written variables

benchislett committed:
Move FileIO::destroy protocol to the end of main()

Instead of closing the file and re-opening each iteration of `evolve`,
we keep it open as long as possible and close when the program terminates,
currently closing at the end of main() in main.cu, and test_all() in unit_test.cu

benchislett committed:
Keep an unordered map of datasets for overwriting

Instead of previously erroring out when a dataset write happens twice
with the same name, we keep track of the datasets and overwrite the data

benchislett committed:
Remove nullification from FileIO::destroy

benchislett committed:
Add writeOutParams

Replace Grid::write and writeOutParam with writeOutParams,
which adds the params as attributes at the root of output.h5
In doing so, getters were added for the private double and integer unordered_maps

benchislett committed:
Add compression props

benchislett committed:
Add general compression to output

benchislett committed:
Add escape in FileIO::writeNd when n=1

benchislett committed:
Move `gstate` into `par` and make it a boolean

benchislett committed:
clean up compiler warnings and unused allocations

benchislett committed:
Remove makefile, as it can no longer be used

benchislett committed:
Move writeOut steps outside wfc loop`

benchislett committed:
Create skeleton structure for reading in wfc

benchislett committed:
Add more to read from hdf5 functionality

Add string/bool map getters,
Add loadParams, loadWfc, and loadA

benchislett committed:
Add iteration counters for imaginary and real-time

benchislett committed:
Remove excess fileio functions

benchislett committed:
Add comments and doc-comments to FileIO changes

benchislett committed:
Update helpfile

benchislett committed:
Move default values to their own map

benchislett committed:
Use conditional storage when reading in attributes

benchislett committed:
Rework energy calculation in evolve()

benchislett committed:
rework iteration variables in evolve()

benchislett committed:
Add writing of energy as a series of attributes

benchislett committed:
Add map in Grid for std::vector<double> values

benchislett committed:
Add write_it check to output of energy

benchislett committed:
Update energy output attribute names

benchislett committed:
Merge branch 'development' into output-to-hdf5

benchislett committed:
Remove unused comments in CMakeLists.txt

benchislett committed:
Remove step_offset parsing

benchislett committed:
Remove all references to step_offset

mlxd commented on src/io/fileIO.cu:
```diff
@@ -0,0 +1,651 @@
+    // Output file
+    H5File *output;
+
+    // Groups
```
Curious if these are freed anywhere? I have seen that some of the pointer types are freed later, but I may be missing it for the rest of these. I wonder if a run through valgrind to check for leaks would be a good idea.

Also wondering if it would make sense to use unique_ptr types instead so that we do not need to manually manage memory safety. Thoughts?

benchislett replied:
Yeah, the groups are missing from the deletion block.
I'll add them for clarity, but we are really only need to destroy the file, to ensure that it is closed.
Since these are never overwritten, they should hang around until program termination and be automatically freed, but I agree that it is better to err on the side of caution.

With regards to `unique_ptr`s, it would be great to use them, but I am using pointers here because they are used everywhere else.
Their types should remain consistent regardless of where they are.
If we do want to use `unique_ptr`s, we should use them everywhere.

That sounds like a bit of an undertaking, but doable.
I'll defer to @leios on making a call here

mlxd replied:
Rewriting to use unique_ptr everywhere would be a bit of an undertaking. Adding the free's to the destroy method is fine by me. I'll have a more detailed look at the rest when I get a chance, but seems all good to me. Nice work.

benchislett replied:
Just to clarify, the `delete` deallocates the pointer and triggers the destructor of the hdf5 objects, which in turn deallocate and close themselves. We don't need to explicitly call `free`

benchislett committed:
Add all FileIO pointers to the deallocator

benchislett committed:
Remove parsing of Afn as file

leios reviewed (commented):
I have been through the PR a few times and I am relatively happy with the changes; however, I have a few notes:

This PR removes the Makefile completely for a pure CMake build, right?

I think I need to write a script to take ASCII data and turn it into an input file for GPUE. This would allow us to read in A's or wfc's generated by other languages.

benchislett commented on src/simulator/evolution.cu:
```diff
@@ -799,8 +740,7 @@ void evolve(Grid &par,
                         cudaCheckError();
-                    }
-                    else{
+                    } else{
```
Totally on purpose. I don't see the point of having an extra line for an else statement; it should be on the same line as the previous closing brace to show that the conditional statement isn't over yet.
Feel free to undo if you really want to, it's not a huge deal

leios replied:
Yeah, I don't care. This is a style thing. I am personally against it, but I am also against ternary operators.

benchislett commented on src/io/fileIO.cu:
```diff
@@ -0,0 +1,670 @@
+
+    // Initialize the data spaces needed for IO
+    void createDataSpaces(int xDim, int yDim, int zDim, int dimnum, int wfc_num) {
+        int rank = 1 + dimnum; // number of components x spatial dimensions
```
One rank for the `wfc_num`. The comment says "number of components x spatial dimensions".
Perhaps it isn't clear enough, but the number of components corresponds to the single rank (`1`) and the spatial dimensions is variable (`dimnum`). Hence, the product of wfc_num and xDim,yDim,zDim gives a rank of 1 + (one of {1, 2, 3})

benchislett committed:
Remove commented inclusions

benchislett commented on include/lattice.h:
```diff
@@ -24,7 +24,7 @@
 #include "node.h"
 #include "edge.h"
-#include "tracker.h"
+//#include "tracker.h"
```
Removed.

benchislett commented:
> This PR removes the Makefile completely for a pure CMake build, right?
Right.

mlxd commented:
I haven't had much time to look at the changes in here, but from what I have seen things look good to me. As James said earlier, ensuring backwards compatibility is a good idea, but an additional file for processing to ASCII data to HDF is possible (something in Python might be quick and easy for this).

As for performance, AFAIK we don't have any regression testing anywhere (CI/CD is hard for CUDA after all, and expensive). But have there been any runs of the original code and this version matching apples to apples simulation performance?

leios commented:
I am getting a roughly 25% speedup. The more IO we do, the faster this is.

I am currently trying to replicate some 3D results, but I am having some trouble. Not sure if this is on my end or if there is something going wrong with the output file. I'm going to look into this more tomorrow.

I think having a script to allow for the user to easily generate the H5 input is essential. I'll also look at this tomorrow. It shouldn't be too hard.

leios committed:
adding quick fix for energy calc / unit test. (#43)

benchislett committed:
Add conditional loading of partial input files

Add two more variables for `par`, `read_file` and `read_a`.
The "old" `read_wfc` has been moved to `read_file`.

When loading an input file, the program will now continue if either of the
datasets is not present, and use the respective default initialization.

benchislett committed:
Merge branch 'feature/output-to-hdf5' of https://github.com/gpue-group/gpue into feature/output-to-hdf5

leios commented:
So I tested inputting an incomplete set of data (that might be created from a separate program) by writing the following python script:

```py
import h5py
import numpy as np

def write_wfc(data_im, data_real, xDim, yDim, zDim, gstate,
              wfc_idx, i, output_filename="../data/output.h5"):
    f = h5py.File(output_filename, "w")
    lines_real = np.loadtxt(data_real)
    lines_im = np.loadtxt(data_im)
    print(len(lines_real))
    wfc_real = np.reshape(lines_real, (xDim,yDim,zDim));
    wfc_im = np.reshape(lines_im, (xDim,yDim, zDim));
    wfc = wfc_real + 1j * wfc_im

    f["/WFC/{}/{}".format("CONST" if gstate else "EV", i)] = wfc

write_wfc("../data_2D_example/wfc_evi_0", "../data_2D_example/wfc_ev_0",
          512, 512, 1, False, 0, 0,
          output_filename="../data_2D_example/output.h5")

```

After running `./gpue -x 512 -y 512 -g 10001 -p 10000 -e 1 -W -r "data_2D_example/output.h5"`, I got the following error:

```
Argument for x is given as 512
Argument for y is given as 512
Argument for Groundsteps is given as 1.000100E+04
Argument for Printout is given as 10000
Argument for EvSteps is given as 1.000000E+00
Writing out
Reading data from file.
Start: Wed Aug 21 15:09:15 2019

Loading attribute gSize with value 65536
Loading attribute plan_1d with value 2
Loading attribute plan_2d with value 1
Loading attribute plan_3d with value 4
Loading attribute plan_dim2 with value 1585669764
Loading attribute plan_dim3 with value 22038
Loading attribute plan_other2d with value 3
Loading attribute zDim with value 1
ERROR: could not find string gstate in Grid::param_bool.
Segmentation fault
```

benchislett committed:
Add gstate to par defaults

benchislett committed:
Fix typo in ds.h

benchislett committed:
Conditionally open groups when loading from file

In the case where some groups are missing from the input file, we
should check if they exist, and load or create the group accordingly.

benchislett committed:
Fix recursion error in fileIO

benchislett committed:
Rename output file to `data.h5`

leios committed:
Merge branch 'development' into hdf5_dev

leios commented:
Ok, I am happy to merge this now. I think everyone has had time to have their say.
